Active Record Migrations > 3. Writing a Migration (#300)
campuscode/rails-guides-pt-BR: pt-BR/active_record_migrations.md

TIP: Dê uma olhada no `--help` dos *generators* (geradores) obter mais detalhes.

Escrevendo uma *Migration*
-------------------

Depois de criar sua *migration* usando um dos geradores, é hora de
começar os trabalhos!

### Criando uma Tabela 

O método `create_table` é um dos mais fundamentais, mas na maioria das vezes,
ele será criado para você usando um gerador de *scaffold* ou *model*. Um uso
típico seria:

```ruby
create_table :products do |t|
  t.string :name
end
```

> Que cria uma tabela `produtos` com uma coluna chamada `nome` (e como discutido

Author: ok

> Que cria uma tabela `produtos` com uma coluna chamada `nome` (e como discutido

Reviewer: > acho que aqui como não estamos modificando o código poderia manter em inglês `products` e `name`. Já traduzimos bastante esses models em outros guias então acho que não teria essa necessidade num "mais avançado" como esse. O que acha? @campuscode/guia-rails-team podem dar uma olhada?

Concordo em manter no inglês e acho que fica mais coerente com o próprio exemplo da migration acima.

> Que cria uma tabela `produtos` com uma coluna chamada `nome` (e como discutido

Reviewer 2: Eu também concordo em manter em inglês.

abaixo, uma coluna implícita `id`.

Por padrão, o `create_table` criará uma chave primária chamada `id`. Você pode mudar
o nome da chave primária com a opção `:primary_key` (não esqueça de
atualizar o *model* correspondente) ou, se você não quer uma chave primária, você
pode passar a opção `id: false`. Se você precisa passar opções específicas do banco de dados 
você pode colocar um fragmento SQL na opção `:option`. Por exemplo:

```ruby
create_table :products, options: "ENGINE=BLACKHOLE" do |t|
  t.string :name, null: false
end
```

acrescentará `ENGINE=BLACKHOLE` à instrução SQL usada para criar a tabela.

Você também pode passar a opção `:comment` com qualquer descrição para a tabela
que serão armazenados no próprio bando de dados e poderão ser visualizados com ferramentas de administração
de bando de dados, como MySQL Workbench ou PgAdmin III. É altamente recomendável especificar

> comentários nas *migrations* para aplicações com grandes bancos de dados, pois ajuda as pessoas

Reviewer: ```suggestion
comentários nas *migrations* para aplicações com grandes bancos de dados, pois ajudam as pessoas
```

a entender o modelo de dados e gerar documentação.
Atualmente, apenas os adaptadores MySQL e PostgreSQL suportam comentários.

> ### Criando uma Tabela Join

Reviewer 3: Esse conceito já apareceu diversas vezes. Acho que já traduzimos para tabela associativa, tabela intermediária, tabela de junção, entre outros.

Acho que podemos deixar em português e inglês esse termo minha dúvida seria o que usar em português, alguma ideia @maugravena e @campuscode/guia-rails-team ?

Reviewer 2: Eu tava traduzindo como `tabela de junção`, mas não sei se tem algum termo melhor, ou se é mais indicado deixar em inglês.

Uma sugestão é deixar com o termo em português e o termo em inglês do lado:

```suggestion
### Criando uma Tabela de Junção (Join Table)
```

Reviewer 3: Eu vejo muito como tabela associativa também mas `tabela de junção` talvez fique mais parecido com o que o Rails faz né.

@tkusuki e @akaninja tem mais alguma ideia também?

Se concordarem a gente pode incluir Tabela de Junção na doc de tradução para futuras referencias.

Author: Acredito que a sugestão de usar tabela de junção está muito boa, ficaria um termo bem próximo da palavra em inglês e de seu significado.

Reviewer: > Eu vejo muito como tabela associativa também mas `tabela de junção` talvez fique mais parecido com o que o Rails faz né.
> 
> @tkusuki e @akaninja tem mais alguma ideia também?
> 
> Se concordarem a gente pode incluir Tabela de Junção na doc de tradução para futuras referencias.

Gosto de junção também, fica bem próximo do termo em inglês. O [Wikipedia](https://pt.wikipedia.org/wiki/Join_(SQL)) também trata como "junção".

> O metódo de *migration* `create_join_table` cria uma tabela *join* HABTM (tem e pertence a

Reviewer 3: ```suggestion
O método de *migration* `create_join_table` cria uma tabela *join* HABTM (tem e pertence a
```

muitos). Um uso comum seria:

```ruby
create_join_table :products, :categories
```

que cria uma tabela `categories_products` com duas colunas chamadas
`category_id` e `product_id`. Essas colunas têm a opção `:null` definida como
`false` por padrão. Isso pode ser substituído, especificando a opção
`:column_options`:

```ruby
create_join_table :products, :categories, column_options: { null: true }
```

Por padrão, o nome da tabela de *join* vem da união dos dois primeiros

> argumentos fornecidos para `create_join_table` em ordem alfabética;

Reviewer 3: ```suggestion
argumentos fornecidos para `create_join_table` em ordem alfabética.
```

> Para customizar o nome da table, forneça uma opção `:table_name`:

Reviewer: ```suggestion
Para customizar o nome da tabela, forneça uma opção `:table_name`:
```

end
```

> ### Mudando tabelas

Reviewer: ```suggestion
### Mudando Tabelas
```

end
```

remove as colunas `description` e `name`, cria uma coluna de string `part_number`
e adiciona um índice nela. Finalmente renomeia a coluna `upccode`.

> ### Mundado Colunas

Reviewer: ```suggestion
### Mudando Colunas
```

Assim como o `remove_column` e `add_column`, o Rails fornece o método de
*migration* `change_column`.

```ruby
change_column :products, :part_number, :text
```

Isso muda a coluna `part_number` na tabela produtos para ser um campo `:text`.
Note que o comando `change_column` é irreversível.

Além do `change_column`, os métodos `change_column_null` e `change_column_default`

> são usados especificamente para alterar uma *not null constraint* e valores

Reviewer 3: O que acha de traduzirmos esse `not null constraint` para algo como `restrição de não ser nulo`, fica ruim?

O que acha @campuscode/guia-rails-team alguém com uma ideia?

Reviewer: > O que acha de traduzirmos esse `not null constraint` para algo como `restrição de não ser nulo`, fica ruim?
> 
> O que acha @campuscode/guia-rails-team alguém com uma ideia?

Concordo.

```

Isso define o campo `:name` em produtos para uma coluna `NOT NULL` e o valor
padrão do campo `:approved` de `true` para `false`.

NOTE: Você também pode escrever a *migration* `change_column_default` acima como
`change_column_default :products, :approved, false`, mas diferente do exemplo
anterior, isso tornaria sua *migration* irreversível.

### Modificadores de Coluna

Modificadores de coluna podem ser aplicados ao criar ou alterar uma coluna:

* `limit`        Define o tamanho máximo dos campos `string/text/binary/integer`.
* `precision`    Define a precisão para os campos `decimal`, representando a
quantidade total de dígitos no número.
* `scale`        Define a escala para os campos `decimal`, representando o
número de digitos após o ponto decimal.
* `polymorphic`  Adiciona uma coluna `type` para a associação `belongs_to`.
* `null`         Autoriza ou não valores `NULL` na coluna.
* `default`      Permite definir um valor padrão na coluna. Note que se você
estiver usando um valor dinâmico (como uma data), o padrão será calculado
apenas na primeira vez (ou seja, na data em que a *migration* é aplicada).
* `comment`      Adiciona um comentário para a coluna.

> Alguns adaptadores podem suportar opções adicionais; consulte a documentação da API de um adaptador espécifico

Reviewer 3: ```suggestion
Alguns adaptadores podem suportar opções adicionais; consulte a documentação da API de um adaptador especifico
```

Author: acredito que aqui o acento está no lugar errado. Estão ficaria específico, correto?

Reviewer 3: Atualizei a sugestion 👍

para maiores informações.

NOTE: `null` e `default` não podem ser especificados via linha de comando.

> ### Foreign Keys (Chaves Estrangeiras)

Reviewer 3: 👏

> Embora não seja necessário, você pode adicionar restrições de foreign key (chave estrageira) para

Reviewer 3: ```suggestion
Embora não seja necessário, você pode adicionar restrições de foreign key (chave estrangeira) para
```

[garantir a integridade referencial](#active-record-and-referential-integrity).

```ruby
add_foreign_key :articles, :authors
```

> Isso adiciona uma nova foreign key (chave estrangeira) à coluna `author_id` da tabela

Reviewer 3: ```suggestion
Isso adiciona uma nova *foreign key* (chave estrangeira) à coluna `author_id` da tabela
```

`articles`. A chave referencia a coluna `id` para a tabela `authors`. Se os
nomes da coluna não puderem ser derivados dos nomes das tabelas, você poderá usar as

> opções `:column` e `:primary_key`.

Reviewer 3: acho que embaixo faltou uma linha

> O Rails irá gerar um nome para cada foreign key (chave estrangeira) começando com

Reviewer 3: ```suggestion
O Rails irá gerar um nome para cada *foreign key* (chave estrangeira) começando com
```

`fk_rails_` seguido por 10 caracteres que são gerados

> deterministicamente a partir do `from_table` e `column`.

Reviewer: ```suggestion
especificamente a partir do `from_table` e `column`.
```

Existe uma opção `:name` para especificar um nome diferente se necessário.

> NOTE: O Active Record suporta apenas foreign keys (chaves estrangeiras) de coluna única. `execute` e

Reviewer 3: ```suggestion
NOTE: O Active Record suporta apenas *foreign keys* (chaves estrangeiras) de coluna única. `execute` e
```

`structure.sql` são obrigados a usar foreign keys (chaves estrangeiras) compostas. Consulte
[Schema Dumping e Você](#schema-dumping-and-you).

> A remoção de uma foreign key (chave estrangeira) também é fácil:

Reviewer: ```suggestion
A remoção de uma *foreign key* (chave estrangeira) também é possível:
```
Recentemente o @odineiramone compartilhou esse [PR no Guides oficial](https://github.com/rails/rails/pull/37946) onde são removidas palavras como "just", "simple" ou "easy" por serem um possível gatinho de frustração para quem ta iniciando. Achei bem legal e poderíamos seguir aqui também. O que acham?

Mais informações aqui https://justsimply.dev/

Author: Achei uma boa ideia, mas não seria melhor aguardar a mudança no guia original? Assim teríamos uma tradução mais próxima do ideal.

Reviewer 3: Acho uma excelente ideia ❤️ ³. Não acho que precisemos esperar o Guides atualizar, no final quando a atualização acontecer teremos que conferir essa linha de qualquer forma. Então não deve fazer muita diferença mudar agora ou depois.

remove_foreign_key :accounts, name: :special_fk_name
```

> ### Quando os Helpers não são Suficientes

Reviewer: ```suggestion
### Quando os Helpers não são suficientes
```

Author: Nas traduções dos títulos sempre sigo as letras maiúsculas quando aparece em algumas palavras, acha que nesse caso ficaria estranho utilizar assim?

Reviewer: > Nas traduções dos títulos sempre sigo as letras maiúsculas quando aparece em algumas palavras, acha que nesse caso ficaria estranho utilizar assim?

Acredito que não, é só :art:

Se os *helpers* fornecidos pelo Active Record não forem suficientes, você poderá usar o método `execute`
para executar SQL arbitrário:

```ruby
Product.connection.execute("UPDATE products SET price = 'free' WHERE 1=1")
```

Para mais detalhes e exemplos de métodos individuais, consulte a documentação da API.
Em particular, a documentação para
[`ActiveRecord::ConnectionAdapters::SchemaStatements`](https://api.rubyonrails.org/classes/ActiveRecord/ConnectionAdapters/SchemaStatements.html)

> (que fornece os métodos disponíveis nos métodos `change`,` up` e `down`),

Reviewer 3: ```suggestion
(que fornece os métodos disponíveis nos métodos `change`, `up` e `down`),
```

* rename_index
* rename_table

`change_table` também é reversível, desde que o bloco não chame `change`,
`change_default` ou `remove`.

`remove_column` é reversível se você fornecer o tipo de coluna como o terceiro
argumento. Forneça também as opções da coluna original, caso contrário, o Rails poderá

> recriar a coluna exatamente ao reverter:

Reviewer 3: ```suggestion
recriar a coluna exata ao reverter:
```

Acho que aqui o exatamente ficou estranho pode dar uma olhada?

Author: concordo que usar *exatamente* ficou estranho, mas acho que *exato* não seria a melhor tradução. Parece que o *exatamente* expressa uma ação que vai acontecer naquele *exato momento*, ou seja, no momento de rever.

..._recriar exatamente quando rever_, seria uma boa tradução?

Reviewer 3: Reli a frase aqui acho que problema é a falta de um não.

seria na verdade algo como *caso contrário, O Rails **não** poderá recriar...*

Author: Pessoal, faz sentido meu raciocínio?

Author: Verdade, vou traduzir assim então :

_o Rails não poderá recriar a coluna exatamente ao reverter:_

pode ser?

Reviewer 3: Parece legal :)

end
```

O uso de `reversible` garantirá que as instruções também sejam executadas na
ordem certa. Se o exemplo anterior de *migration* for revertido,

> o bloco `down` será executado depois da coluna `home_page_url` for removida e

Reviewer: ```suggestion
o bloco `down` será executado depois que a coluna `home_page_url` for removida e
```

logo antes da tabela `distributors` for apagada.

Às vezes sua *migration* fará algo que é simplesmente irreversível;
por exemplo, pode destruir alguns dados. Em alguns casos, você pode levantar o
`ActiveRecord::IrreversibleMigration` no seu bloco `down`. Se alguém tentar
reverter sua *migration*, uma mensagem de erro será exibida dizendo que isso
não pode ser feito.

### Usando os métodos `up`/`down`

Você também pode usar o estilo antigo de *migration* usando os métodos `up` e `down`
em vez do método `change`.
O método `up` deve descrever a transformação que você deseja fazer no seu
*schema*, e o método `down` da sua *migration* deve reverter as

> transformações feitas pelo método `up`. Em outras palavas, o *schema* do bando de dados

Reviewer 3: ```suggestion
transformações feitas pelo método `up`. Em outras palavas, o *schema* do banco de dados
```

end
```

> Se sua *migration* é irreversível, você deve levandatar um

Reviewer 3: ```suggestion
Se sua *migration* é irreversível, você deve dar `raise` num
```

Acho que na doc não é um comando de código mas ficaria mais fácil como código, o que acha?

@campuscode/guia-rails-team

Author: concordo
